tests/gtestmusig2.c

```diff
@@ -0,0 +1,416 @@
+#include <gtest/gtest.h>
+extern "C" {
+#include "../src/libmusig2.h"
+#ifndef NR_SIGS
+#define NR_SIGS 3
+#endif
+#include "config.h"
+
+
+
+TEST (musig2, not_enough_signatures) {
+
+    secp256k1_context *ctx = secp256k1_context_create(SECP256K1_CONTEXT_VERIFY);
+
+
+    int i, j, k, l;
+    int ind;
+    secp256k1_pubkey pk_list[NR_SIGNERS];    // Signers' public key list
+    secp256k1_pubkey batch_list[NR_SIGNERS * V * NR_MESSAGES];   // Stores the batches of signers
+    musig2_context_sig mcs_list[NR_SIGNERS]; // Array that holds NR_SIGNERS musig2_context_sig
+
+    int err;
+
+    /**** Initialization ****/
+    for (i = 0; i < NR_SIGNERS; i++) {
+        /* Generate a keypair for the signer and get batch commitments. */
+        err = musig2_init_signer(&mcs_list[i], ctx, NR_MESSAGES);
+        ASSERT_EQ(err, 1);
+
+        /* Store the public key of the signer in pk_list */
+        err = secp256k1_keypair_pub(ctx, &pk_list[i], &mcs_list[i].keypair);
+        ASSERT_EQ(err, 1);
+
+        /* Store the batch commitments of the signer in batch_list */
+        l = 0; // the index of the signer's commitment list.
+        for (k = 0; k < NR_MESSAGES; k++) {
+            for (j = 0; j < V; j++, l++) {
+                ind = (NR_SIGNERS * V * k) + (j * NR_SIGNERS) + i; // the index for the list of collected batches.
+                err = secp256k1_keypair_pub(ctx, &batch_list[ind], mcs_list[i].comm_list[l]);
+                ASSERT_EQ(err, 1);
+            }
+        }
+    }
+
+    /**** Aggregate the public keys and batch commitments for each signer ****/
+    int cnt = 0;
+    for (i = 0; i < NR_SIGNERS; i++)
+        cnt += musig2_aggregate_pubkey(mcs_list[i].mc, pk_list, NR_SIGNERS);
+    ASSERT_EQ(cnt, NR_SIGNERS);
+
+    cnt = 0;
+    for (i = 0; i < NR_SIGNERS; i++)
+        cnt += musig2_aggregate_R(&mcs_list[i], batch_list);
+    ASSERT_EQ(cnt, NR_SIGNERS);
+
+
+    /**** Signature ****/
+    musig2_partial_signatures mps1[NR_SIGS];
+
+    for (i = 0; i < NR_SIGS; i++) {
+        /* Generate the partial signatures */
+        err = musig2_sign(&mcs_list[i], MSG_1, MSG_1_LEN, mps1[i].sig);
+        ASSERT_EQ(err, 1);
+        memcpy(mps1[i].R.data, mcs_list[i].mc->aggr_R.data, PK_BYTES);
```

Comment:
We should re-write the `musig2_sign` function. Currently it takes as input `musig2_partial_signatures.sig`, but it should take the whole structure. In this way we avoid having to do a memcpy after each `musig2_sign` call.